### PR TITLE
Distance metric integration for vamana, and refactoring of distance metrics

### PR DESCRIPTION
**Pull request by cainamisir**

### CODE REVIEW

File: src/include/index/vamana_index.h [hunk 1]
```diff
@@ -179,14 +180,16 @@ class vamana_index {
       size_t num_nodes,
       size_t l_build,
       size_t r_max_degree,
-      std::optional<TemporalPolicy> temporal_policy = std::nullopt)
+      std::optional<TemporalPolicy> temporal_policy = std::nullopt,
+      DistanceMetric distance_metric = DistanceMetric::L2)
```
Comment (NikolaosPapailiou) [inline src/include/index/vamana_index.h:184]: Can we pass the distance function here and instantiate all the different values in dispatch_tables?

File: apis/python/test/test_distance_metrics.py [hunk 2]
```diff
@@ -109,41 +383,69 @@ def test_wrong_type_with_distance_metric(tmp_path):
...
+    # now create index with L2 and print results
+
+    index_uri = os.path.join(tmp_path, "sift10k_flat_L2")
+    index2 = ingest(
+        index_type="IVF_FLAT",
+        index_uri=index_uri,
+        input_vectors=input_vectors,
+    )
+
+    distances2, ids2 = index2.query(query_vectors, k=5)
```
Comment (jparismorgan) [inline apis/python/test/test_distance_metrics.py:440]: Can we check that these are equal to the expected values? And same above.

File: apis/python/test/test_distance_metrics.py [hunk 3]
```diff
@@ -1,32 +1,306 @@
...
+
+    update_vectors = np.empty([5], dtype=object)
+    update_vectors[0] = np.array([1, 0, 0, 0], dtype=vector_type)
+    update_vectors[1] = np.array([1, 1, 0, 0], dtype=vector_type)
+    update_vectors[2] = np.array([32, 41, 30, 0], dtype=vector_type)
+    update_vectors[3] = np.array([1, 5, 3, 0], dtype=vector_type)
+    update_vectors[4] = np.array([4, 4, 4, 0], dtype=vector_type)
+    index.update_batch(vectors=update_vectors, external_ids=np.array([0, 1, 2, 3, 4]))
+
+    expected_distances = np.array([0.500000, 0.292893, 0.142262, 0.239361, 0.133975])
```
[the remark this answers is not in the document]
Comment (cainamisir) [reply apis/python/test/test_distance_metrics.py:269]: Oh I just did it by hand since it was such a small test case

File: src/include/api/vamana_index.h [hunk 4]
```diff
@@ -506,38 +521,111 @@ class IndexVamana {
...
+
+  {{TILEDB_FLOAT32, TILEDB_UINT32, TILEDB_UINT64, DistanceMetric::L2}, [](size_t num_vectors, uint64_t l_build, uint64_t r_max_degree, std::optional<TemporalPolicy> temporal_policy, DistanceMetric distance_metric) { return std::make_unique<index_impl<vamana_index<float,   uint32_t, uint64_t, sum_of_squares_distance>>>(num_vectors, l_build, r_max_degree, temporal_policy, distance_metric); }},
+  {{TILEDB_FLOAT32, TILEDB_UINT32, TILEDB_UINT64, DistanceMetric::INNER_PRODUCT}, [](size_t num_vectors, uint64_t l_build, uint64_t r_max_degree, std::optional<TemporalPolicy> temporal_policy, DistanceMetric distance_metric) { return std::make_unique<index_impl<vamana_index<float,   uint32_t, uint64_t, inner_product_distance>>>(num_vectors, l_build, r_max_degree, temporal_policy, distance_metric); }},
+  {{TILEDB_FLOAT32, TILEDB_UINT32, TILEDB_UINT64, DistanceMetric::COSINE}, [](size_t num_vectors, uint64_t l_build, uint64_t r_max_degree, std::optional<TemporalPolicy> temporal_policy, DistanceMetric distance_metric) { return std::make_unique<index_impl<vamana_index<float,   uint32_t, uint64_t, cosine_distance>>>(num_vectors, l_build, r_max_degree, temporal_policy, distance_metric); }},
```
Comment (jparismorgan) [inline src/include/api/vamana_index.h:551]: So I had thought Vamana only supports L2 and Cosine? If that's the case, can we remove `DistanceMetric::INNER_PRODUCT` from here? If that's not the case, can we add unit tests for `DistanceMetric::INNER_PRODUCT`?

File: src/include/api/vamana_index.h [hunk 5]
```diff
@@ -422,6 +436,7 @@ class IndexVamana {
       // @todo using index_type = size_t;
       auto dtype = vectors.feature_type();
 
+      // print distance function typr
```
Comment (jparismorgan) [inline src/include/api/vamana_index.h:439]: (Minor) Let's remove this, and also the places where we add newlines to different places in the code, please

File: src/include/api/vamana_index.h [hunk 6]
```diff
@@ -292,13 +303,15 @@ class IndexVamana {
...
+      tiledb_datatype_t* distance_metric) {
     using metadata_element =
         std::tuple<std::string, tiledb_datatype_t*, tiledb_datatype_t>;
     std::vector<metadata_element> metadata{
         {"feature_datatype", feature_datatype, TILEDB_UINT32},
-        {"id_datatype", id_datatype, TILEDB_UINT32}};
-
+        {"id_datatype", id_datatype, TILEDB_UINT32},
+        {"distance_metric", distance_metric, TILEDB_UINT32}};
+    // add distance metric here
```
Comment (jparismorgan) [inline src/include/api/vamana_index.h:314]: (Minor) Let's remove, please

File: src/include/test/unit_api_vamana_index.cc [hunk 7]
```diff
@@ -901,3 +902,58 @@ TEST_CASE("write and load index with timestamps", "[api_vamana_index]") {
         std::vector<uint64_t>{100}.begin()));
   }
 }
+TEST_CASE("vamana cosine distance", "[api_vamana_index]") {
```
Comment (jparismorgan) [inline src/include/test/unit_api_vamana_index.cc:905]: (Minor) Could you also add something to this test case where we open up a new index by URI and verify that we still use Cosine, and that the scores are correct? That will test the `read_types()` changes.

File: apis/python/src/tiledb/vector_search/vamana_index.py [hunk 8]
```diff
@@ -168,11 +168,14 @@ def create(
...
         )
-    ctx = vspy.Ctx(config)
-    if distance_metric != vspy.DistanceMetric.L2:
+    if (
+        distance_metric != vspy.DistanceMetric.L2
+        and distance_metric != vspy.DistanceMetric.COSINE
+    ):
         raise ValueError(
-            f"Distance metric {distance_metric} is not supported in VAMANA"
+            f"Distance metric {distance_metric} is not supported in IVF_PQ"
```
Comment (jparismorgan) [inline apis/python/src/tiledb/vector_search/vamana_index.py:176]:
```suggestion
            f"Distance metric {distance_metric} is not supported in VAMANA"
```

File: apis/python/src/tiledb/vector_search/index.py [hunk 9]
```diff
@@ -67,6 +67,7 @@ def __init__(
             raise ValueError(
                 f"Invalid distance metric in metadata: {self.group.meta.get('distance_metric')}."
             )
+
```
Comment (jparismorgan) [inline apis/python/src/tiledb/vector_search/index.py:70]: (Minor) Let's undo this change, please

File: src/include/api/vamana_index.h [hunk 10]
```diff
@@ -90,7 +91,6 @@ class IndexVamana {
       const std::optional<IndexOptions>& config = std::nullopt) {
     feature_datatype_ = TILEDB_ANY;
     id_datatype_ = TILEDB_UINT32;
-
```
Comment (jparismorgan) [inline src/include/api/vamana_index.h:93]: (Minor) Let's undo this change, please

File: src/include/api/vamana_index.h [hunk 11]
```diff
@@ -131,10 +128,19 @@ class IndexVamana {
...
-    read_types(ctx, group_uri, &feature_datatype_, &id_datatype_);
-
+    tiledb_datatype_t distance_metric_raw;
+    read_types(
+        ctx,
+        group_uri,
+        &feature_datatype_,
+        &id_datatype_,
+        &distance_metric_raw);
+    distance_metric_ = static_cast<DistanceMetric>(distance_metric_raw);
```
Comment (jparismorgan) [inline src/include/api/vamana_index.h:138]: (Minor) This is okay, but why not just have `read_types()` return `DistanceMetric`? Seems cleaner?

File: src/include/api/vamana_index.h [hunk 12]
```diff
@@ -385,6 +397,7 @@ class IndexVamana {
           extents(training_set)[1]};
 
       using id_type = typename T::id_type;
+
```
Comment (jparismorgan) [inline src/include/api/vamana_index.h:400]: (Minor) Let's undo this change, please

**Review by jparismorgan — APPROVED** [review]
Nice!

File: src/include/api/vamana_index.h (hunk 11 above)
Comment (cainamisir) [reply src/include/api/vamana_index.h:138]: we would have to move the distance_metric_raw logic inside the read_types and i thought it would be different than the pattern we already have - i can definitely move it there, it just seemed cleaner

Comment (jparismorgan) [reply src/include/api/vamana_index.h:139]: Either way, your choice

**Review by NikolaosPapailiou — APPROVED** [review]
